Fulfill the capstone project and its acceptance criteria

Ilyeo opened the pull request:
Quick Info
---
[Preview link](https://tiny-souffle-1dc5ef.netlify.app/)

Ilyeo committed:
Update dependencies to latest versions

Ilyeo committed:
Add `.env` file to the git ignore list

Ilyeo committed:
Add `Header` component using tdd

Ilyeo committed:
Add `Footer` component using tdd

Ilyeo committed:
Add `Layout` component

Ilyeo committed:
Add `msw` library as development dependency

Ilyeo committed:
Add mock service worker custom setup

As a complement to our suite of tests, we need a server to respond to
our requests.

This server just response mocked data for our tests, we can add more
custom mocked responses throught `handler.js` file.

Ilyeo committed:
Update `ReactDom.render` by `createRoot`

In favor to use React 18 this change is required for more info please
follow this link below.

https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-client-rendering-apis

Ilyeo committed:
Add `user-event` library

Ilyeo committed:
Add abstraction to fetch an apod from nasa services

`getApod` it's a promise that returns the astronomy picture of the day
(a.k.a. APOD) given a date.

Ilyeo committed:
Add `Apod` and `ApodError` componets using tdd

Ilyeo committed:
Add tests to `Home` component

Ilyeo committed:
Add `@nextui-org/react` library

Ilyeo committed:
Add ui improvements

`@nextui-org/react` provide us with awesome built-in components ready
for use.

Ilyeo committed:
Add safe guard to return with impossible date

An sample of impossible date is 30-02-2022.

PacoMojica commented on src/components/Apod/Apod.test.jsx:
```diff
@@ -0,0 +1,51 @@
+  "url":"https://apod.nasa.gov/apod/image/2204/Apollo-16-station-10crop1110.jpg",
+}
+
+const setup = () => render(<Apod {...mock} />)
```
this could be executed in a `beforeEach`, makes the test easier to understand and reduces the duplicated code

[https://jestjs.io/docs/setup-teardown](https://jestjs.io/docs/setup-teardown)

Ilyeo replied:
thanks for the advice i'll do it

PacoMojica commented on src/components/Apod/Apod.test.jsx:
```diff
@@ -0,0 +1,51 @@
+    expect(apodDate).toHaveAttribute('datetime')
+  })
+
+  it('should display apod image', () => {
```
the assertions of the the test cases could be moved to a single test because they are testing that the component renders the expected data

Ilyeo replied:
you're right

PacoMojica commented on src/components/Footer/Footer.test.jsx:
```diff
@@ -0,0 +1,24 @@
+const setup = () => render(<Footer />)
+
+describe('Footer', () => {
+  it('should be in the document', () => {
+    setup()
+
+    const footer = screen.getByRole('contentinfo')
+
+    expect(footer).toBeInTheDocument()
+  })
+
+  it('should display "Project created during Wizeline Academy React Testing Bootcamp"', () => {
```
these tests cases could also be merged

PacoMojica commented on src/components/Header/Header.test.jsx:
```diff
@@ -0,0 +1,24 @@
+const setup = () => render(<Header />)
+
+describe('Header', () => {
+  it('should be in the document', () => {
+    setup()
+
+    const header = screen.getByRole('banner')
+
+    expect(header).toBeInTheDocument()
+  })
+
+  it('should display the project title', () => {
```
same here, the test cases could be merged

PacoMojica commented on src/pages/Home.test.jsx:
```diff
@@ -0,0 +1,76 @@
+import { server } from '../mocks/server.js'
+import { Home } from './Home'
+
+const setup = () => render(<Home />)
```
this `setup` could be executed in a `beforeEach` as well

PacoMojica commented on src/pages/Home.test.jsx:
```diff
@@ -0,0 +1,76 @@
+
+    expect(datePicker).toHaveValue(today)
+
+    await waitFor(() => screen.getByRole('heading'))
+
+    const apodTitle = screen.getByRole('heading')
```
there is a selector called `findBy...` which can be used instead of the `waitFor + getBy`, something like this:
```js
await waitFor(() => screen.getByRole('heading'))
const apodTitle = screen.getByRole('heading')
```

can be changed to:

```js
const apodTitle = await screen.findByRole('heading')
```

Ilyeo replied:
super cool! i did not know this

PacoMojica commented on src/mocks/handlers.js:
```diff
@@ -0,0 +1,21 @@
+import { rest } from 'msw'
+
+export const handlers = [
+  rest.get(process.env.NASA_APOD_URL, (req, res, ctx) => {
+    req.url.searchParams.set('api_key', process.env.NASA_API_KEY)
```
the env var for the API key could be hardcoded and can be anything, I think that having it as a env var is not required for the test

Ilyeo replied:
😅

Ilyeo committed:
Addressing mentorship comments

Ilyeo commented:
[@PacoMojica](https://github.com/PacoMojica) thanks for all your comments

Ilyeo committed:
Add `useApod` custom hook

This hook receives an object of probable keys(in this case only a
`date`) to execute a request to NASA service.
